### PR TITLE
fix(codex): dedupe replayed usage

### LINKED ISSUE #1434: Significant accounting issues for codex
### What happened?

I know this will get auto-closed but just wanted to let you know there are currently major accounting issues with codex for the past month @ryoppippi. I think it has something to do with subagents having all historical transcripts leading up to them get counted. For some days it seems like my usage is even 10x or 20x higher counting than expected. It might have something to do with recent codex updates, I am not too sure.

I'm looking into it now because it is deteriorating my personal tooling. If I succeed I can make a PR. Just wanted to make an issue to give a heads up.

There are other issues too for this. They were all auto-closed:

- #1349 (June 21, 2026)
- #1337 (June 18, 2026)
- #1288 (June 11, 2026)
- #950 (April 19, 2026)

### Steps to reproduce

N/A

### Expected behavior

_No response_

### Version

20.0.17

### PR DESCRIPTION
**Pull request by MilesCranmer**
Codex fork and subagent logs replay parent `token_count` history with rewritten timestamps. The existing same-second heuristic misses replay bursts spanning multiple seconds and nested forks, causing the copied history to be aggregated again. Codex can also re-emit `last_token_usage` snapshots without advancing `total_token_usage`.

Match forked session event prefixes against immutable parent streams before aggregation. This handles replay bursts spanning multiple seconds and nested parents without mutating ancestry needed by children. When the referenced parent log is unavailable, retain the existing same-second fallback.

Skip last-token snapshots when cumulative totals do not advance, preserving session-aware and alias-aware report deduplication.

Addresses #950, #988, #1288, #1337, and #1349.

@ryoppippi this fixes #1434

I dogfooded this locally on a frozen 16 GB Codex history. The corrected report matched the sampled root-session totals in Codex’s `state_5.sqlite`.

---
<a href="https://app.blacksmith.sh/ccusage/codesmith/ccusage/pr/1435"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with Codesmith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1786465095&installation_id=139163553&pr_number=1435&repository=ccusage%2Fccusage&return_to=https%3A%2F%2Fgithub.com%2Fccusage%2Fccusage%2Fpull%2F1435&signature=e82071299e4e1be075f1c095d9bbaa2ed681ff1d59ee469aec223d1f2a663174"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>/codesmith</code> with what you need. Autofix is disabled.</sup>

---
## Summary by cubic
Deduplicates replayed Codex usage by matching each fork’s leading events to its immutable parent stream, bounded at the fork (RFC3339 or epoch). Falls back to same‑second detection when the parent can’t anchor the replay and skips non‑advancing snapshots so session/alias reports stay correct.

- **Bug Fixes**
  - Per‑session replay plan; drop only the replayed prefix; bound at RFC3339/epoch; skip self‑parent sessions; keep real post‑fork child events even if equal; works single‑threaded and in parallel.
  - Parser does two‑phase replay filtering: try parent‑prefix match, then same‑second fallback if the first event doesn’t match; detect the fallback second from raw token‑count records before cumulative suppression to handle missing parents, multi‑second, mid‑stream, and nested replays.
  - Ignore `last_token_usage` when cumulative totals don’t advance. Fixes #1434; addresses #950, #988, #1288, #1337, #1349.

- **Refactors**
  - Group aggregation inputs into `CodexAggregateRun`; read session metadata on all workers; resolve duplicate session IDs to the first file; normalize numeric fork timestamps via shared event timestamp parsing.
  - Tests: reuse shared replay fixture builders.

<sup>Written for commit 8b4078c9dee99cd8449e785dfe1b6709549bc853. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/ccusage/ccusage/pull/1435?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved Codex usage reporting by generating and applying per-session “parent usage” context via a replay plan during session parsing (consistent for sequential and parallel runs).
  * Reworked replay detection to use prefix-based filtering with timestamp/second alignment, avoiding incorrect replay skipping and duplicate token usage.
  * Refined event usage selection (favoring totals when available, otherwise using last-token usage).

* **Tests**
  * Expanded coverage for replay skipping correctness, missing-parent handling, nested replays, and fork/next-event preservation, including `skips_forked_parent_prefix_rewritten_across_seconds`.

### CODE REVIEW

**Comment by MilesCranmer** [comment]
@coderabbitai review
@cubic-dev-ai review

File: rust/crates/ccusage/src/adapter/codex/parser.rs [hunk 1]
```diff
@@ -267,8 +177,12 @@ fn visit_codex_session_entry(
     }
     let info = payload.info.as_ref();
     let total_usage = info.and_then(|info| info.total_token_usage.as_ref().copied());
+    let cumulative_advanced = total_usage
+        .as_ref()
+        .is_none_or(|total_usage| previous_totals.as_ref() != Some(total_usage));
     let raw_usage = info
         .and_then(|info| info.last_token_usage.as_ref().copied())
+        .filter(|_| cumulative_advanced)
```
[the remark this answers is not in the document]
Comment (MilesCranmer) [reply rust/crates/ccusage/src/adapter/codex/parser.rs:278]: I don't think this is P1, it's kind of an edge case if someone manually deletes parent logs for some reason. But yeah we can fix it.

[the remark this answers is not in the document]
Comment (MilesCranmer) [reply rust/crates/ccusage/src/adapter/codex/parser.rs:278]: Fixed.

**Comment by MilesCranmer** [comment]
I fixed the AI suggestion although since the PR is closed it won't show up until re-opened.

**Comment by MilesCranmer** [comment]
@ryoppippi could you please re-open this? Thanks

**Comment by axisrow** [comment]
Investigated this on my local Codex history (1,524 rollout files, **122 with `forked_from_id` and 97 with `thread_spawn`**) and read through the diff. This is a real improvement over #1369 and I'd like to see it merged, with one suggestion.

**Why this is stronger than the current `#1369` heuristic.** The same-second detection in `parser.rs` only fires when ≥2 `token_count` events share an identical wall-clock second. Codex rewrites replayed timestamps with millisecond offsets that regularly span 2+ seconds, so `#1369` misses them — which is exactly the inflated `gpt-5`/"phantom" usage I (and others in #1434) still see on v20.0.17. Matching parent→child via `session_meta` (`forked_from_id` / `source/subagent/thread_spawn/parent_thread_id`) and skipping the prefix by full delta equality is the right fix. Running both the parent stream and the child traversal through the same `visit_codex_session_file` path (so both sides emit derived deltas symmetrically) is the detail that makes value-based matching safe — nice.

**Code review — looks correct:**
- `cumulative_advanced` filter (`parser.rs`): correctly only suppresses `last_token_usage` when `total_token_usage` is unchanged; `previous_totals` is tracked on every available total, and the first event is never skipped. ✔️
- Memory: `usage_by_parent` holds full streams only for files that are actual parents (~40 bytes/event), not all files. Acceptable. ✔️
- Threading: `&CodexReplayPlan` is read-only and scoped correctly; `read_parent_usage`'s `thread::scope` is sound. ✔️

**One suggestion before merge — pin down the false-positive case.** Because the prefix skip matches on *values*, a non-replay child whose first real delta happens to equal the parent's first delta would lose that one event before the match breaks (`matching_replay = false`). The severity is low (a single event; a multi-event false match would require identical token-generation sequences), but under-counting is worse than over-counting, and right now no test proves false skips don't happen — only that true skips do. A test like this would close it and document the invariant:

- parent: first delta `{in:100, out:5}`
- an unrelated child (no `forked_from_id`) whose first real event is also `{in:100, out:5}`, followed by a distinct second event
- assert **both** child events survive

With that test added, this is a +1 to merge from me. @ryoppippi could you reopen? The auto-close caught a new contributor; the underlying issue (#1434) is actively biting people.

**Comment by Nebu1eto** [comment]
Whilst I appreciate that leaving a '+1' comment on a given issue or PR is generally frowned upon, but I strongly recommend this PR. There is currently a bug within ccusage that erroneously calculates Codex usage (#1434); I've built from this PR's branch and performed validation using my own Codex data.

**Review by ryoppippi — APPROVED** [review]
LGTM

**Comment by ryoppippi** [comment]
this is just a format issue

**Comment by ryoppippi** [comment]
@pullfrog review it
@codesmith review it
@cubic review it
@coderabbitai review it


and @codesmith. fix ci!

**Comment by ryoppippi** [comment]
@MilesCranmer why we need a new github workflow(release native? what is that?)

it should not be included the fix of our logic.

plz fix this

**Comment by MilesCranmer** [comment]
> @MilesCranmer why we need a new github workflow(release native? what is that?)
> 
> it should not be included the fix of our logic.
> 
> plz fix this

@ryoppippi That wasn't me, I think it was one of your AI agents? 

Will fix

**Comment by MilesCranmer** [comment]
Oh wait no that was totally me! (Sorry AI...) Was my fault, my local coding agent pushed that to the branch when I was trying to make a custom release I could use.

Anyways, will fix.

**Comment by MilesCranmer** [comment]
@ryoppippi ok ready

File: rust/crates/ccusage/src/adapter/codex/parser.rs [hunk 2]
```diff
@@ -154,7 +125,25 @@ pub(super) fn visit_codex_session_file(
...
+    let mut visit_filtered = |event: CodexTokenUsageEvent| {
+        if matching_replay {
+            let usage = CodexRawUsage {
+                input_tokens: event.input_tokens,
+                cached_input_tokens: event.cached_input_tokens,
+                output_tokens: event.output_tokens,
+                reasoning_output_tokens: event.reasoning_output_tokens,
+                total_tokens: event.total_tokens,
+            };
+            if replayed_prefix.and_then(|prefix| prefix.get(replay_index)) == Some(&usage) {
```
[the remark this answers is not in the document]
Comment (MilesCranmer) [reply rust/crates/ccusage/src/adapter/codex/parser.rs]: Good catch — this was a real false positive when the parent continued after the fork. I now bound the replay prefix to parent events at or before the child session metadata fork timestamp and added a regression where the child first real usage equals the parent next post-fork event. Fixed in 3208776.

File: rust/crates/ccusage/src/adapter/codex/replay.rs [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+        return CodexSessionMetadata::default();
+    }
+    let Ok(value) = serde_json::from_str::<Value>(&line) else {
+        return CodexSessionMetadata::default();
+    };
+    let payload = (value.get("type").and_then(Value::as_str) == Some("session_meta"))
+        .then_some(value.get("payload"))
+        .flatten();
+    CodexSessionMetadata {
+        timestamp: value
```
[the remark this answers is not in the document]
Comment (MilesCranmer) [reply rust/crates/ccusage/src/adapter/codex/replay.rs]: The other agent told me to add that part 😭

**Comment by ryoppippi** [comment]
oh my god.
thank you. i'll take a look tmrw!

@MilesCranmer

File: rust/crates/ccusage/src/adapter/codex/replay.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ryoppippi) [reply rust/crates/ccusage/src/adapter/codex/replay.rs]: @MilesCranmer so sorry.... do you wanna keep it or not

[the remark this answers is not in the document]
Comment (MilesCranmer) [reply rust/crates/ccusage/src/adapter/codex/replay.rs]: Yeah we can keep it, it's no big deal

**Comment by ryoppippi** [comment]
Reviewed this and pushed two follow-up commits to the branch. Thanks for the thorough test coverage — it made the edge cases easy to probe.

**1. The missing-parent fallback could destroy a parent stream** (`6278ec33`)

The fallback stored its first-second prefix in `usage_by_parent` keyed by the child's own path. When that file was also another session's parent — a missing grandparent, a parent, and a child — the insert replaced the parent's full usage stream with only its first replayed second, so the grandchild stopped matching after two events and counted the rest of the replayed history again:

```
actual:   [("01-parent", 300), ("01-parent", 400), ("02-child", 300), ("02-child", 400), ("02-child", 500)]
expected: [("01-parent", 300), ("01-parent", 400), ("02-child", 500)]
```

`ParentReplay::path` is now optional and `replay_prefix` returns an empty slice when the parent log is unavailable, so a child path is never used as a parent key. Covered by `keeps_full_parent_stream_when_the_parent_itself_replayed_a_missing_session`.

**2. A found-but-unmatched parent lost the same-second fallback** (`6278ec33`)

Prefix matching starts at the parent's first event, and the same-second fallback only applied when the parent log was missing. So when a replay does not line up with the start of the parent stream, dedupe was skipped entirely — `main` still handled those files:

```
actual:   [.., ("02-child", 200), ("02-child", 300), ("02-child", 400)]
expected: [.., ("02-child", 400)]
```

The same-second heuristic is now a second state of the replay match in the parser, entered when the very first event fails to match. A partial match still means the mismatch is the fork point, so the bound added in `32087768` is unchanged. Covered by `falls_back_to_rewritten_second_when_the_replay_starts_mid_parent_stream`.

**3. Smaller items** (`6278ec33`)

- A session listing itself in `forked_from_id` would have matched its own stream and lost every event it recorded; that parent is now ignored.
- Session metadata is read on all workers instead of serially on the main thread. On a local 1880-file history the plan phase drops from ~105 ms to ~78 ms warm, and cold runs no longer serialize one open per file.
- Duplicate session ids resolve to the first file so the plan does not depend on source ordering.
- The fork timestamp is parsed from RFC3339 only; Codex writes `session_meta.timestamp` as a string, and the numeric branch guessed seconds against milliseconds with an unexplained threshold.

**4. CI was red on clippy** (`5b552a9b`)

Threading the replay plan through the aggregation helpers pushed `aggregate_file` to eight parameters, which fails `clippy::too_many_arguments` under `-D warnings`. The inputs that are constant for one run now travel in a `CodexAggregateRun` struct.

One thing I did not change: `cumulative_advanced` suppresses `last_token_usage` for every session, not just forks, and assumes `total_token_usage` is cumulative. That matches the existing `subtract_codex_raw_usage` path, so it is not a new assumption, but it is worth keeping in mind alongside #1458.

**Comment by ryoppippi** [comment]
@coderabbitai full review

**Comment by ryoppippi** [comment]
Verified on real data: 1880 local Codex session files, `ccusage codex --json`, release builds of `021f6631` (this PR before my follow-ups) versus `7bdc5375`.

| | totalTokens | costUSD |
| --- | --- | --- |
| before | 23,785,447,263 | 12,929.34 |
| after | 23,771,689,798 | 12,927.03 |

Exactly one daily row changed (2026-07-13, -13,757,465 tokens of replayed history that was still being counted twice); every other day is identical, so the follow-ups only remove duplicates and do not drop real usage.

**Comment by ryoppippi** [comment]
Addressed the cubic nit in `8b4078c9`: `keeps_child_usage_matching_parent_event_after_fork` now uses the module-level `replay_metadata` / `replay_token_count` builders instead of its own copies.

**Comment by han-cheng6** [comment]
@MilesCranmer Well done! Now my case look good now.

```
--- before upgrading(remove unrelated):

$ date; ccusage-codex
2026年 5月18日 星期一 16时16分12秒 CST

 WARN  Fetching latest model pricing from LiteLLM...                                                                                                                        @ccusage/codex 16:16:23

ℹ Loaded pricing for 2716 models                                                                                                                                           @ccusage/codex 16:16:24

 WARN  Pricing not found for model codex-auto-review; defaulting to zero-cost pricing.                                                                                      @ccusage/codex 16:16:24


 ╭──────────────────────────────────────────────────────────────╮
 │                                                              │
 │  Codex Token Usage Report - Daily (Timezone: Asia/Shanghai)  │
 │                                                              │
 ╰──────────────────────────────────────────────────────────────╯

┌──────────────┬──────────────────────────────┬──────────────┬─────────────┬────────────┬────────────────┬────────────────┬─────────────┐
│ Date         │ Models                       │        Input │      Output │  Reasoning │     Cache Read │   Total Tokens │  Cost (USD) │
├──────────────┼──────────────────────────────┼──────────────┼─────────────┼────────────┼────────────────┼────────────────┼─────────────┤
│ May 12, 2026 │ - codex-auto-review          │   35,607,673 │   2,913,722 │    713,857 │    569,439,104 │    607,960,499 │     $508.84 │
│              │ - gpt-5.5                    │              │             │            │                │                │             │
└──────────────┴──────────────────────────────┴──────────────┴─────────────┴────────────┴────────────────┴────────────────┴─────────────┘

--- after upgrading(remove unrelated, version 20.0.19):

$ date; npx ccusage@latest codex
2026年 8月10日 星期一 20时13分40秒 CST

╭────────────────────────────────────────────╮
│                                            │
│      Codex Token Usage Report - Daily      │
│                                            │
╰────────────────────────────────────────────╯

┌────────────┬───────────────┬──────────────┬─────────────┬─────────────┬────────────────┬─────────────────┬─────────────┐
│ Date       │ Models        │        Input │      Output │   Reasoning │     Cache Read │    Total Tokens │  Cost (USD) │
├────────────┼───────────────┼──────────────┼─────────────┼─────────────┼────────────────┼─────────────────┼─────────────┤
│ 2026-05-12 │ - gpt-5.5     │   12,741,647 │   1,332,183 │     260,691 │    194,860,160 │     208,933,990 │     $201.10 │
└────────────┴───────────────┴──────────────┴─────────────┴─────────────┴────────────────┴─────────────────┴─────────────┘
```

**Comment by ryoppippi** [comment]
Historical audit: this pull request was auto-closed by the legacy contributor gate. That closure did not assess technical importance.

Audit result: **resolved**. A later merged change or the current `main` implementation covers this request. This PR is kept for history and does not need to be revived.
